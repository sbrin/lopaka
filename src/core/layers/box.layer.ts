import {RectangleLayer} from './rectangle.layer';

export class BoxLayer extends RectangleLayer {
    protected type: ELayerType = 'box';
    protected fill: true;
}
