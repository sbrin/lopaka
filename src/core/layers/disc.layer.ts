import {CircleLayer} from './circle.layer';

export class DiscLayer extends CircleLayer {
    protected type: ELayerType = 'disc';
    protected fill: boolean = true;
}
