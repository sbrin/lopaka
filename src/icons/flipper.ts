import I125_10px from '../../img/flipper/125_10px.png';
import Alert_9x8 from '../../img/flipper/Alert_9x8.png';
import ArrowC_1_36x36 from '../../img/flipper/ArrowC_1_36x36.png';
import ArrowDownEmpty_14x15 from '../../img/flipper/ArrowDownEmpty_14x15.png';
import ArrowDownFilled_14x15 from '../../img/flipper/ArrowDownFilled_14x15.png';
import ArrowUpEmpty_14x15 from '../../img/flipper/ArrowUpEmpty_14x15.png';
import ArrowUpFilled_14x15 from '../../img/flipper/ArrowUpFilled_14x15.png';
import Attention_5x8 from '../../img/flipper/Attention_5x8.png';
import back_10px from '../../img/flipper/back_10px.png';
import badusb_10px from '../../img/flipper/badusb_10px.png';
import Battery_16x16 from '../../img/flipper/Battery_16x16.png';
import Battery_26x8 from '../../img/flipper/Battery_26x8.png';
import Ble_connected_15x15 from '../../img/flipper/Ble_connected_15x15.png';
import Ble_disconnected_15x15 from '../../img/flipper/Ble_disconnected_15x15.png';
import Bluetooth_Connected_16x8 from '../../img/flipper/Bluetooth_Connected_16x8.png';
import Bluetooth_Idle_5x8 from '../../img/flipper/Bluetooth_Idle_5x8.png';
import Button_18x18 from '../../img/flipper/Button_18x18.png';
import ButtonCenter_7x7 from '../../img/flipper/ButtonCenter_7x7.png';
import ButtonDown_7x4 from '../../img/flipper/ButtonDown_7x4.png';
import ButtonLeft_4x7 from '../../img/flipper/ButtonLeft_4x7.png';
import ButtonLeftSmall_3x5 from '../../img/flipper/ButtonLeftSmall_3x5.png';
import ButtonRight_4x7 from '../../img/flipper/ButtonRight_4x7.png';
import ButtonRightSmall_3x5 from '../../img/flipper/ButtonRightSmall_3x5.png';
import ButtonUp_7x4 from '../../img/flipper/ButtonUp_7x4.png';
import Charging_lightning_9x10 from '../../img/flipper/Charging-lightning_9x10.png';
import Charging_lightning_mask_9x10 from '../../img/flipper/Charging-lightning_mask_9x10.png';
import Clock_18x18 from '../../img/flipper/Clock_18x18.png';
import CoolHi_25x27 from '../../img/flipper/CoolHi_25x27.png';
import CoolHi_hvr_25x27 from '../../img/flipper/CoolHi_hvr_25x27.png';
import CoolLo_25x27 from '../../img/flipper/CoolLo_25x27.png';
import CoolLo_hvr_25x27 from '../../img/flipper/CoolLo_hvr_25x27.png';
import Dehumidify_25x27 from '../../img/flipper/Dehumidify_25x27.png';
import Dehumidify_hvr_25x27 from '../../img/flipper/Dehumidify_hvr_25x27.png';
import Detailed_chip_17x13 from '../../img/flipper/Detailed_chip_17x13.png';
import dir_10px from '../../img/flipper/dir_10px.png';
import Down_25x27 from '../../img/flipper/Down_25x27.png';
import Down_hvr_25x27 from '../../img/flipper/Down_hvr_25x27.png';
import Error_18x18 from '../../img/flipper/Error_18x18.png';
import EviSmile1_18x21 from '../../img/flipper/EviSmile1_18x21.png';
import EviSmile2_18x21 from '../../img/flipper/EviSmile2_18x21.png';
import EviWaiting1_18x21 from '../../img/flipper/EviWaiting1_18x21.png';
import EviWaiting2_18x21 from '../../img/flipper/EviWaiting2_18x21.png';
import FaceCharging_29x14 from '../../img/flipper/FaceCharging_29x14.png';
import FaceConfused_29x14 from '../../img/flipper/FaceConfused_29x14.png';
import FaceNopower_29x14 from '../../img/flipper/FaceNopower_29x14.png';
import FaceNormal_29x14 from '../../img/flipper/FaceNormal_29x14.png';
import GameMode_11x8 from '../../img/flipper/GameMode_11x8.png';
import Health_16x16 from '../../img/flipper/Health_16x16.png';
import HeatHi_25x27 from '../../img/flipper/HeatHi_25x27.png';
import HeatHi_hvr_25x27 from '../../img/flipper/HeatHi_hvr_25x27.png';
import HeatLo_25x27 from '../../img/flipper/HeatLo_25x27.png';
import HeatLo_hvr_25x27 from '../../img/flipper/HeatLo_hvr_25x27.png';
import Hidden_window_9x8 from '../../img/flipper/Hidden_window_9x8.png';
import ibutt_10px from '../../img/flipper/ibutt_10px.png';
import InfraredArrowDown_4x8 from '../../img/flipper/InfraredArrowDown_4x8.png';
import InfraredArrowUp_4x8 from '../../img/flipper/InfraredArrowUp_4x8.png';
import ir_10px from '../../img/flipper/ir_10px.png';
import KeyBackspace_16x9 from '../../img/flipper/KeyBackspace_16x9.png';
import KeyBackspaceSelected_16x9 from '../../img/flipper/KeyBackspaceSelected_16x9.png';
import Keychain_39x36 from '../../img/flipper/Keychain_39x36.png';
import KeySave_24x11 from '../../img/flipper/KeySave_24x11.png';
import KeySaveSelected_24x11 from '../../img/flipper/KeySaveSelected_24x11.png';
import Left_mouse_icon_9x9 from '../../img/flipper/Left_mouse_icon_9x9.png';
import loading_10px from '../../img/flipper/loading_10px.png';
import Lock_7x8 from '../../img/flipper/Lock_7x8.png';
import Lock_8x8 from '../../img/flipper/Lock_8x8.png';
import Medium_chip_22x21 from '../../img/flipper/Medium-chip-22x21.png';
import MHz_25x11 from '../../img/flipper/MHz_25x11.png';
import Modern_reader_18x34 from '../../img/flipper/Modern_reader_18x34.png';
import Move_flipper_26x39 from '../../img/flipper/Move_flipper_26x39.png';
import music_10px from '../../img/flipper/music_10px.png';
import Mute_25x27 from '../../img/flipper/Mute_25x27.png';
import Mute_hvr_25x27 from '../../img/flipper/Mute_hvr_25x27.png';
import Nfc_10px from '../../img/flipper/Nfc_10px.png';
import Off_25x27 from '../../img/flipper/Off_25x27.png';
import Off_hvr_25x27 from '../../img/flipper/Off_hvr_25x27.png';
import Ok_btn_9x9 from '../../img/flipper/Ok_btn_9x9.png';
import Ok_btn_pressed_13x13 from '../../img/flipper/Ok_btn_pressed_13x13.png';
import Percent_10x14 from '../../img/flipper/Percent_10x14.png';
import Pin_arrow_down_7x9 from '../../img/flipper/Pin_arrow_down_7x9.png';
import Pin_arrow_left_9x7 from '../../img/flipper/Pin_arrow_left_9x7.png';
import Pin_arrow_right_9x7 from '../../img/flipper/Pin_arrow_right_9x7.png';
import Pin_arrow_up_7x9 from '../../img/flipper/Pin_arrow_up_7x9.png';
import Pin_attention_dpad_29x29 from '../../img/flipper/Pin_attention_dpad_29x29.png';
import Pin_back_arrow_10x8 from '../../img/flipper/Pin_back_arrow_10x8.png';
import Pin_back_full_40x8 from '../../img/flipper/Pin_back_full_40x8.png';
import Pin_pointer_5x3 from '../../img/flipper/Pin_pointer_5x3.png';
import Pin_star_7x7 from '../../img/flipper/Pin_star_7x7.png';
import Power_25x27 from '../../img/flipper/Power_25x27.png';
import Power_hvr_25x27 from '../../img/flipper/Power_hvr_25x27.png';
import Pressed_Button_13x13 from '../../img/flipper/Pressed_Button_13x13.png';
import Quest_7x8 from '../../img/flipper/Quest_7x8.png';
import Release_arrow_18x15 from '../../img/flipper/Release_arrow_18x15.png';
import Restoring_38x32 from '../../img/flipper/Restoring_38x32.png';
import RFIDBigChip_37x36 from '../../img/flipper/RFIDBigChip_37x36.png';
import Right_mouse_icon_9x9 from '../../img/flipper/Right_mouse_icon_9x9.png';
import SDcardFail_11x8 from '../../img/flipper/SDcardFail_11x8.png';
import SDcardMounted_11x8 from '../../img/flipper/SDcardMounted_11x8.png';
import SDQuestion_35x43 from '../../img/flipper/SDQuestion_35x43.png';
import SmallArrowDown_3x5 from '../../img/flipper/SmallArrowDown_3x5.png';
import SmallArrowDown_4x7 from '../../img/flipper/SmallArrowDown_4x7.png';
import SmallArrowUp_3x5 from '../../img/flipper/SmallArrowUp_3x5.png';
import SmallArrowUp_4x7 from '../../img/flipper/SmallArrowUp_4x7.png';
import Smile_18x18 from '../../img/flipper/Smile_18x18.png';
import sub1_10px from '../../img/flipper/sub1_10px.png';
import Tap_reader_36x38 from '../../img/flipper/Tap_reader_36x38.png';
import Temperature_16x16 from '../../img/flipper/Temperature_16x16.png';
import u2f_10px from '../../img/flipper/u2f_10px.png';
import unknown_10px from '../../img/flipper/unknown_10px.png';
import Unlock_7x8 from '../../img/flipper/Unlock_7x8.png';
import Up_25x27 from '../../img/flipper/Up_25x27.png';
import Up_hvr_25x27 from '../../img/flipper/Up_hvr_25x27.png';
import update_10px from '../../img/flipper/update_10px.png';
import Updating_32x40 from '../../img/flipper/Updating_32x40.png';
import UsbTree_48x22 from '../../img/flipper/UsbTree_48x22.png';
import Vol_down_25x27 from '../../img/flipper/Vol_down_25x27.png';
import Vol_down_hvr_25x27 from '../../img/flipper/Vol_down_hvr_25x27.png';
import Vol_up_25x27 from '../../img/flipper/Vol_up_25x27.png';
import Vol_up_hvr_25x27 from '../../img/flipper/Vol_up_hvr_25x27.png';
import Voldwn_6x6 from '../../img/flipper/Voldwn_6x6.png';
import Voltage_16x16 from '../../img/flipper/Voltage_16x16.png';
import Volup_8x6 from '../../img/flipper/Volup_8x6.png';
import Warning_30x23 from '../../img/flipper/Warning_30x23.png';

export default {
    I125_10px,
    Alert_9x8,
    ArrowC_1_36x36,
    ArrowDownEmpty_14x15,
    ArrowDownFilled_14x15,
    ArrowUpEmpty_14x15,
    ArrowUpFilled_14x15,
    Attention_5x8,
    back_10px,
    badusb_10px,
    Battery_16x16,
    Battery_26x8,
    Ble_connected_15x15,
    Ble_disconnected_15x15,
    Bluetooth_Connected_16x8,
    Bluetooth_Idle_5x8,
    Button_18x18,
    ButtonCenter_7x7,
    ButtonDown_7x4,
    ButtonLeft_4x7,
    ButtonLeftSmall_3x5,
    ButtonRight_4x7,
    ButtonRightSmall_3x5,
    ButtonUp_7x4,
    Charging_lightning_9x10,
    Charging_lightning_mask_9x10,
    Clock_18x18,
    CoolHi_25x27,
    CoolHi_hvr_25x27,
    CoolLo_25x27,
    CoolLo_hvr_25x27,
    Dehumidify_25x27,
    Dehumidify_hvr_25x27,
    Detailed_chip_17x13,
    dir_10px,
    Down_25x27,
    Down_hvr_25x27,
    Error_18x18,
    EviSmile1_18x21,
    EviSmile2_18x21,
    EviWaiting1_18x21,
    EviWaiting2_18x21,
    FaceCharging_29x14,
    FaceConfused_29x14,
    FaceNopower_29x14,
    FaceNormal_29x14,
    GameMode_11x8,
    Health_16x16,
    HeatHi_25x27,
    HeatHi_hvr_25x27,
    HeatLo_25x27,
    HeatLo_hvr_25x27,
    Hidden_window_9x8,
    ibutt_10px,
    InfraredArrowDown_4x8,
    InfraredArrowUp_4x8,
    ir_10px,
    KeyBackspace_16x9,
    KeyBackspaceSelected_16x9,
    Keychain_39x36,
    KeySave_24x11,
    KeySaveSelected_24x11,
    Left_mouse_icon_9x9,
    loading_10px,
    Lock_7x8,
    Lock_8x8,
    Medium_chip_22x21,
    MHz_25x11,
    Modern_reader_18x34,
    Move_flipper_26x39,
    music_10px,
    Mute_25x27,
    Mute_hvr_25x27,
    Nfc_10px,
    Off_25x27,
    Off_hvr_25x27,
    Ok_btn_9x9,
    Ok_btn_pressed_13x13,
    Percent_10x14,
    Pin_arrow_down_7x9,
    Pin_arrow_left_9x7,
    Pin_arrow_right_9x7,
    Pin_arrow_up_7x9,
    Pin_attention_dpad_29x29,
    Pin_back_arrow_10x8,
    Pin_back_full_40x8,
    Pin_pointer_5x3,
    Pin_star_7x7,
    Power_25x27,
    Power_hvr_25x27,
    Pressed_Button_13x13,
    Quest_7x8,
    Release_arrow_18x15,
    Restoring_38x32,
    RFIDBigChip_37x36,
    Right_mouse_icon_9x9,
    SDcardFail_11x8,
    SDcardMounted_11x8,
    SDQuestion_35x43,
    SmallArrowDown_3x5,
    SmallArrowDown_4x7,
    SmallArrowUp_3x5,
    SmallArrowUp_4x7,
    Smile_18x18,
    sub1_10px,
    Tap_reader_36x38,
    Temperature_16x16,
    u2f_10px,
    unknown_10px,
    Unlock_7x8,
    Up_25x27,
    Up_hvr_25x27,
    update_10px,
    Updating_32x40,
    UsbTree_48x22,
    Vol_down_25x27,
    Vol_down_hvr_25x27,
    Vol_up_25x27,
    Vol_up_hvr_25x27,
    Voldwn_6x6,
    Voltage_16x16,
    Volup_8x6,
    Warning_30x23
};
