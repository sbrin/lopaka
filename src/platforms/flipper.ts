import {AbstractImageLayer} from '../core/layers/abstract-image.layer';
import {AbstractLayer} from '../core/layers/abstract.layer';
import {FontFormat} from '../draw/fonts/font';
import {bdfFonts} from '../draw/fonts/fontTypes';
import {imgDataToXBMP, toCppVariableName} from '../utils';
import {Platform} from './platform';
import defaultTemplate from './templates/flipper/default.pug';

export class FlipperPlatform extends Platform {
    public static id = 'flipper';
    protected name = 'Flipper Zero';
    protected description = 'Flipper Zero';
    protected fonts: TPlatformFont[] = [
        {
            title: 'FontPrimary',
            name: 'FontPrimary',
            file: bdfFonts.find((f) => f.name === 'helvB08').file,
            format: FontFormat.FORMAT_BDF
        },
        {
            title: 'FontSecondary',
            name: 'FontSecondary',
            file: bdfFonts.find((f) => f.name === 'haxrcorp4089').file,
            format: FontFormat.FORMAT_BDF
        },
        {
            title: 'FontKeyboard',
            name: 'FontKeyboard',
            file: bdfFonts.find((f) => f.name === 'profont11').file,
            format: FontFormat.FORMAT_BDF
        },
        {
            title: 'FontBigNumbers',
            name: 'FontBigNumbers',
            file: bdfFonts.find((f) => f.name === 'profont22').file,
            format: FontFormat.FORMAT_BDF
        }
    ];

    constructor() {
        super();
        this.features.hasInvertedColors = false;
        this.features.defaultColor = '#000000';
        this.features.screenBgColor = '#ff8200';
    }

    protected templates = {
        Default: {
            template: defaultTemplate,
            settings: {}
        }
    };

    generateSourceCode(layers: AbstractLayer[], ctx?: OffscreenCanvasRenderingContext2D): string {
        const declarations: {type: string; data: any}[] = [];
        const xbmps = [];
        const xbmpsNames = [];
        const layerData = layers
            .sort((a: AbstractLayer, b: AbstractLayer) => a.index - b.index)
            .map((layer) => {
                const props = layer.properties;
                if (layer instanceof AbstractImageLayer && !layer.imageName) {
                    const XBMP = imgDataToXBMP(layer.data, 0, 0, layer.size.x, layer.size.y).join(',');
                    if (xbmps.includes(XBMP)) {
                        props.imageName = xbmpsNames[xbmps.indexOf(XBMP)];
                    } else {
                        const varName = `image_${xbmps.length + 1}_bits`;
                        declarations.push({
                            type: 'bitmap',
                            data: {
                                name: varName,
                                value: XBMP
                            }
                        });
                        xbmps.push(XBMP);
                        xbmpsNames.push(varName);
                        props.imageName = varName;
                    }
                }
                return props;
            });
        const source = this.templates[this.currentTemplate].template({
            declarations,
            layers: layerData,
            settings: Object.assign({}, this.settings, this.templates[this.currentTemplate].settings)
        });
        return source;
    }
}
