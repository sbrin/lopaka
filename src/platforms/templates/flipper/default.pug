each decl in declarations 
    if decl.type == 'bitmap'
        | static const uint8_t !{decl.data.name}[] = {!{decl.data.value}};
        |
    else 
        | !{decl}
        |
| canvas_set_bitmap_mode(canvas, true);
|
if settings.wrap
    | void draw(Canvas* canvas) {
- var font = ''
each layer in layers 
    case layer.type
        when 'dot'
            | @!{layer.id};canvas_draw_dot(canvas, !{layer.x}, !{layer.y});
        when 'line'
            | @!{layer.id};canvas_draw_line(canvas, !{layer.x1}, !{layer.y1}, !{layer.x2}, !{layer.y2});
        when 'rect'
            - var func = layer.fill ? 'canvas_draw_box' : 'canvas_draw_frame'
            | @!{layer.id};!{func}(canvas, !{layer.x}, !{layer.y}, !{layer.w}, !{layer.h});
        when 'circle'
            - var func = layer.fill ? 'canvas_draw_disc' : 'canvas_draw_circle'
            | @!{layer.id};!{func}(canvas, !{layer.x}, !{layer.y}, !{layer.r});
        when 'string'
            if layer.font != font
                - font = layer.font
                | canvas_set_font(canvas, !{layer.font});
                |
            | @!{layer.id};canvas_draw_str(canvas, !{layer.x}, !{layer.y}, "!{layer.text}");
        when 'paint'
            | @!{layer.id};canvas_draw_xbm(canvas, !{layer.x}, !{layer.y}, !{layer.w}, !{layer.h}, !{layer.imageName});
        when 'icon'
            | @!{layer.id};canvas_draw_icon(canvas, !{layer.x}, !{layer.y},  &I_!{layer.imageName});
        default 
            |
            | // Unknown layer type !{layer.type}
            |
    |
    |
if settings.wrap
    | }