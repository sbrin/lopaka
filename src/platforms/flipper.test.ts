import {describe, expect, it} from 'vitest';
import {layersMock} from './layers.mock';
import {FlipperPlatform} from './flipper';

describe('Flipper zero platform', () => {
    it('generating source code', () => {
        const platform = new FlipperPlatform();
        expect(platform.generateSourceCode(layersMock)).toMatchSnapshot();
    });
});
