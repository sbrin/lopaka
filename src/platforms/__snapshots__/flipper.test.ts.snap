// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flipper zero platform > generating source code 1`] = `
"static const uint8_t image_1_bits[] = {0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00};
canvas_set_bitmap_mode(canvas, true);
@veqtjp8jf9ln6isyfz;canvas_draw_box(canvas, 107, 46, 18, 15);
@utlxpqvhdxnln6it5pw;canvas_draw_box(canvas, 97, 2, 27, 11);
@eu6bpe0xpvvln6itc2e;canvas_draw_box(canvas, 106, 19, 19, 18);
canvas_set_font(canvas, helvB08_tr);
@s3m4u762qpln6jvm90;canvas_draw_str(canvas, 3, 11, \\"Text\\");
@b11j6xs5t46ln6jw96h;canvas_draw_line(canvas, 2, 16, 68, 16);
@6n8rac0nxpcln6jwdwr;canvas_draw_dot(canvas, 62, 4);
@ema2hee85mln6jweht;canvas_draw_dot(canvas, 67, 8);
@j1ol1lo5nkln6jwfmg;canvas_draw_dot(canvas, 72, 12);
@6uf3k6wj82ln6jwi73;canvas_draw_circle(canvas, 20, 41, 12);

// Unknown layer type ellipse

@aibkefmz3cln6jwlit;canvas_draw_disc(canvas, 17, 43, 3);
@g7pk1wfbqqkln6jwxby;canvas_draw_frame(canvas, 44, 29, 44, 28);

// Unknown layer type ellipse

@rrsjcz4oz2ln6jx4y4;canvas_draw_xbm(canvas, 0, 0, 128, 64, image_1_bits);
"
`;
